3scaleAdapter/pkg/threescale/proxy_conf.go

```diff
@@ -0,0 +1,108 @@
+package threescale
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	sysC "github.com/3scale/3scale-porta-go-client/client"
+	"github.com/3scale/istio-integration/3scaleAdapter/config"
+
+	"istio.io/istio/pkg/log"
+)
+
+// DefaultCacheTTL - Default time to wait before purging expired items from the cache
+const DefaultCacheTTL = time.Duration(time.Minute * 1)
+const cacheKey = "%s-%s"
+
+type proxyStore struct {
+	element sysC.ProxyConfigElement
+	setAt   time.Time
+}
+
+// proxyConfigController offers a way to cache proxy configurations from
+// 3scale system.
+type proxyConfigController struct {
+	limit int
+	ttl   time.Duration
+	cache map[string]proxyStore
+	mutex sync.RWMutex
+}
+
+func newProxyConfigController(cacheTTL time.Duration, limit int) *proxyConfigController {
+	pcc := &proxyConfigController{
+		limit: limit,
+		ttl:   cacheTTL,
+		cache: make(map[string]proxyStore, limit),
+	}
+	go pcc.flushCache(cacheTTL)
+	return pcc
+}
+
+func (pc *proxyConfigController) get(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	var conf sysC.ProxyConfigElement
+	var err error
+
+	cacheKey := getCacheKeyFromCfg(cfg)
+
+	pc.mutex.RLock()
+	e, ok := pc.cache[cacheKey]
+	pc.mutex.RUnlock()
+
+	if !ok {
+		conf, err = pc.getFromRemote(cfg, c)
+	} else {
+		log.Debugf("proxy config fetched from cache for service id %s", cfg.ServiceId)
+		conf = e.element
+	}
+
+	return conf, err
+}
+
+func (pc *proxyConfigController) getFromRemote(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	log.Debugf("proxy config for service id %s not cached - fetching from 3scale", cfg.ServiceId)
+	pce, err := c.GetLatestProxyConfig(cfg.AccessToken, cfg.ServiceId, "production")
+	if err == nil {
+		go pc.set(cfg, pce)
+	}
+	return pce, err
+}
+
+func (pc *proxyConfigController) set(cfg *config.Params, e sysC.ProxyConfigElement) {
+	cacheKey := getCacheKeyFromCfg(cfg)
+
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+	if len(pc.cache) < pc.limit {
+		pc.cache[cacheKey] = proxyStore{
+			setAt:   time.Now(),
+			element: e,
+		}
+	}
+}
+
+func getCacheKeyFromCfg(cfg *config.Params) string {
+	return fmt.Sprintf(cacheKey, cfg.SystemUrl, cfg.ServiceId)
+}
+
+func (pc *proxyConfigController) flushCache(t time.Duration) {
+	for {
+		<-time.After(t)
+		now := time.Now()
+
+		pc.mutex.Lock()
```

Comment:
This is quite a coarse mutex that has the potential to slow down a lot other lookups with O(N). I have little experience with Go's standard map structure, but can we iterate over it without breaking anything, even if we don't get the full set of entries, when the map is potentially being written to? This would help in taking the lock only when actually needing to delete an entry.

Reply:
Yes, understood. Let me take a look into that